### PR TITLE
Rewrite implementation for input and output description

### PR DESCRIPTION
**Pull request by bszmelcz**
### Details:
- subtask for Serialization Implementation Improvement
 - rewrite implementation for input and output description for on_adapter function for adapter with void type
### Tickets:
 - 46976

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 1]
```diff
@@ -151,100 +151,106 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+    void output_descriptions_on_adapter(ngraph::ValueAccessor<void>& adapter) {
+        const auto& a = ngraph::as_type<ngraph::AttributeAdapter<std::vector<std::shared_ptr
+                            <ngraph::op::util::SubGraphOp::OutputDescription>>>>(&adapter);
+
+        std::vector<std::string> result_mapping = map_type_from_body(m_xml_node.parent(), "Result");
+        std::vector<std::string> parameter_mapping = map_type_from_body(m_xml_node.parent(), "Parameter");
+
+        NGRAPH_CHECK(!result_mapping.empty(), "No results found in body Function.");
+
+
```
Comment (sdurawa) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:201]: remove empty line

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 2]
```diff
@@ -151,100 +151,106 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+            pugi::xml_node exec_output = port_map.append_child("output");
+            exec_output.append_attribute("external_port_id").set_value("-1");
+            exec_output.append_attribute("internal_layer_id").set_value(result_mapping[a->get().body_condition_output_idx].c_str());
+            exec_output.append_attribute("purpose").set_value("execution_condition");
+        }
+    }
+
+    void on_adapter(const std::string& name, ngraph::ValueAccessor<void>& adapter) override {
+        if (m_xml_node.parent().child("body")) {
+            if (name == "input_descriptions")
```
Comment (sdurawa) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:248]: we shall use incoming adapter type instead its name.

**Review by sdurawa — APPROVED** [review]

**Comment by sdurawa** [comment]
@bszmelcz, please switch to regular PR.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 3]
```diff
@@ -151,100 +151,100 @@ class XmlSerializer : public ngraph::AttributeVisitor {
         return output;
     }
 
-    void on_adapter(const std::string& name,
-                    ngraph::ValueAccessor<void>& adapter) override {
-        if (m_xml_node.parent().child("body")) {
-            // parameters and results from body are required for port_map attributes serialization
-            std::vector<std::string> parameter_mapping = map_type_from_body(m_xml_node.parent(), "Parameter");
-            std::vector<std::string> result_mapping = map_type_from_body(m_xml_node.parent(), "Result");
+    void input_descriptions_on_adapter(const std::vector<std::shared_ptr<
```
Comment (jdanieck) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:154]: input_descriptions_on_adapter() should be private member function

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 4]
```diff
@@ -151,100 +151,100 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
-                pugi::xml_node port_map = m_xml_node.parent().child("port_map");
-                NGRAPH_CHECK(port_map, "port_map section not found, purpose attribute cannot be added.");
-
-                if (a->get().current_iteration_input_idx != -1) {
-                    pugi::xml_node iter_input = port_map.append_child("input");
-                    iter_input.append_attribute("external_port_id").set_value("-1");
-                    iter_input.append_attribute("internal_layer_id").set_value(parameter_mapping[a->get().current_iteration_input_idx].c_str());
-                    iter_input.append_attribute("purpose").set_value("current_iteration");
-                }
+    void output_descriptions_on_adapter(const std::vector<std::shared_ptr<
```
Comment (jdanieck) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:189]: output_descriptions_on_adapter() should be private member function.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 5]
```diff
@@ -151,100 +151,100 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+                output.prepend_attribute("axis").set_value(concat_output->m_axis);
+                output.append_attribute("start").set_value(concat_output->m_start);
+                output.append_attribute("end").set_value(concat_output->m_end);
+                output.append_attribute("stride").set_value(concat_output->m_stride);
+                output.append_attribute("part_size").set_value(concat_output->m_part_size);
             }
+        }
+    }
+
+    void special_body_ports_on_adapter(const ngraph::op::v5::Loop::SpecialBodyPorts& special_body_ports) {
```
Comment (jdanieck) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:216]: special_body_ports_on_adapter() should be private member function

Comment (jdanieck) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:216]: Is there a reason to use specifically ngraph::op::v5  ?

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 6]
```diff
@@ -151,100 +151,100 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+            pugi::xml_node exec_output = port_map.append_child("output");
+            exec_output.append_attribute("external_port_id").set_value("-1");
+            exec_output.append_attribute("internal_layer_id").set_value(result_mapping[special_body_ports.body_condition_output_idx].c_str());
+            exec_output.append_attribute("purpose").set_value("execution_condition");
+        }
+    }
+
+    void on_adapter(const std::string& name, ngraph::ValueAccessor<void>& adapter) override {
+        if (m_xml_node.parent().child("body")) {
+            if (const auto& a = ngraph::as_type<ngraph::AttributeAdapter<std::vector<std::shared_ptr
```
Comment (jdanieck) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:240]: Please add curly braces to if, like `if () {} else {}`.

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 7]
```diff
@@ -151,100 +151,100 @@ class XmlSerializer : public ngraph::AttributeVisitor {
...
+                output.append_attribute("end").set_value(concat_output->m_end);
+                output.append_attribute("stride").set_value(concat_output->m_stride);
+                output.append_attribute("part_size").set_value(concat_output->m_part_size);
             }
+        }
+    }
+
+    void special_body_ports_on_adapter(const ngraph::op::v5::Loop::SpecialBodyPorts& special_body_ports) {
+        std::vector<std::string> result_mapping = map_type_from_body(m_xml_node.parent(), "Result");
+        std::vector<std::string> parameter_mapping = map_type_from_body(m_xml_node.parent(), "Parameter");
```
Comment (jdanieck) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:218]: `result_mapping` & `parameter_mapping` should be computed in `on_adapter<void>` as before and passed as a reference to new helper functions, now this code is duplicated unnecessarily

**Review by jdanieck — CHANGES_REQUESTED** [review]

File: inference-engine/src/transformations/src/transformations/serialize.cpp [hunk 8]
```diff
@@ -125,6 +125,87 @@ class XmlSerializer : public ngraph::AttributeVisitor {
         return join(adapter.get());
     }
 
+    void input_descriptions_on_adapter(const std::vector<std::shared_ptr<
+                                        ngraph::op::util::SubGraphOp::InputDescription>>& input_descriptions,
+                                        std::vector<std::string> parameter_mapping,
+                                        std::vector<std::string> result_mapping,
```
Comment (jdanieck) [inline inference-engine/src/transformations/src/transformations/serialize.cpp:131]: Looks like we could pass `result_mapping` & `parameter_mapping` by const&. The same comment for other helpers.

**Review by jdanieck — COMMENTED** [review]
Please move also map_type_from_body() to private scope. I know that you didn't touch it but it seems it's not need to be public so we can fix it now.

**Review by jdanieck — APPROVED** [review]
LGTM in advance, please fix the last comments and we are good : )

File: inference-engine/src/transformations/src/transformations/serialize.cpp (hunk 8 above)
Comment (bszmelcz) [reply inference-engine/src/transformations/src/transformations/serialize.cpp:131]: Sure, will do that in a second

**Review by ilyachur — APPROVED** [review]

**Pull request merged**